p2p/subscriber_metrics.go

```diff
@@ -0,0 +1,130 @@
+package p2p
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+const (
+	statusKey    = "status"
+	statusAccept = "accept"
+	statusIgnore = "ignore"
+	statusReject = "reject"
+)
+
+type subscriberMetrics struct {
+	messageNumInst  metric.Int64Counter
+	messageSizeInst metric.Int64Histogram
+
+	messageTimeLastMu sync.Mutex
+	messageTimeLast   time.Time
+	messageTimeInst   metric.Float64Histogram
+
+	subscriptionNum     atomic.Int64
+	subscriptionNumInst metric.Int64ObservableGauge
+	subscriptionNumReg  metric.Registration
```

Comment:
You will not need to store those if callback declared and registered inside metrics constructor. So those values would stay local to constractor and will be captured by callback closure.

Reply:
It's used in Close to do unregister.

Reply:
I wonder if unregister even necessary on shutdown 🤔  It seems to be doing fine without it. 
Gauge is still needed to be stored for sure.

Reply:
It may work without it, but it must clear resources properly. The API asks for it, and it does not hurt anyway.